smart_open/s3.py

```diff
@@ -700,20 +740,181 @@ def to_boto3(self, resource):
     #
     # Internal methods.
     #
-    def _read_from_buffer(self, size=-1):
-        """Remove at most size bytes from our buffer and return them."""
-        size = size if size >= 0 else len(self._buffer)
-        part = self._buffer.read(size)
-        self._current_pos += len(part)
-        return part
 
     def _fill_buffer(self, size=-1):
-        size = max(size, self._buffer._chunk_size)
-        while len(self._buffer) < size and not self._eof:
-            bytes_read = self._buffer.fill(self._raw_reader)
-            if bytes_read == 0:
-                logger.debug('%s: reached EOF while filling buffer', self)
-                self._eof = True
+        # check if existing body range is sufficient to satisfy this request, if not close it so that it re-opens
+        # with an appropriate range.
+        is_stream_limit_before_eof = self._body is not None \
+            and size < 0 \
+            and self._stream_range_to < self._content_length - 1
+        is_request_beyond_stream_limit = self._body is not None \
+            and size > 0 \
+            and self._stream_range_to is not None \
+            and self._stream_range_to - self._position < size
+        if is_stream_limit_before_eof or is_request_beyond_stream_limit:
+            self._body.close()
+            self._body = self._stream_range_from = self._stream_range_to = None
+
+        # open the HTTP request if it's not open already
+        if self._body is None:
+            start = self._position + len(self._buffer)
+            stop = None if size < 0 else \
+                start + size - 1 if self._content_length is None else \
+                start + max(size, self._stream_range) - 1
+            self._open_body(start=start, stop=stop)
+
+        b = [self._stream_from_body(size)]
+        bytes_read = self._buffer.fill(b)
+        if bytes_read == 0:
+            logger.debug('%s: reached EOF while filling buffer', self)
+            self._eof = True
+
+    def _raw_read(self, size=-1):
+        """Internal read from the continuous connection with the remote peer without considering buffering."""
+
+        # If we can figure out that we've read past the EOF, then we can save
+        # an extra API call.
+        reached_eof = True if self._content_length is not None and self._position >= self._content_length else False
```

Comment:
Ternary operator.

Also, here's a question: why do we spend so much effort handling the case when the content length is None? I think we ever reach a point where we need it and don't have it, we should just query it.

What do you think?